packages/core/src/zip/__tests__/select.js

```diff
@@ -0,0 +1,238 @@
+'use strict'
+
+import I from 'immutable'
+import R from 'ramda'
+
+import { elementZipper } from '../'
+import { childrenProperty, isElement, isOfKind } from '../../data'
+
+import {
+  isStringArray,
+  isLocationArray,
+  child,
+  ancestors,
+  descendants,
+  children,
+  propEq,
+  select,
+  ofKind,
+} from '../select'
+
+const zipper = data => elementZipper({})(data)
+
+const data = {
+  kind: ['navigation', 'tabs'],
+  [childrenProperty]: ['tabs', 'settings'],
+  title: 'App',
+  tabs: [
+    {
+      kind: ['tab'],
+      title: 'Murder Bot',
+    },
+    {
+      kind: ['tab'],
+      title: 'Dr Mensah',
+      [childrenProperty]: ['elements', 'deeplink'],
+      elements: [
+        {
+          kind: ['element'],
+          title: 'robot',
+        },
+        {
+          kind: ['element'],
+          title: 'human',
+        },
+      ],
+      deeplink: {
+        kind: ['link'],
+        title: 'goodreads',
+      },
+    },
+  ],
+  settings: {
+    kind: ['settings'],
+    title: 'Martha Wells',
+  },
+}
+
+describe('zip.select', () => {
+  describe('helpers', () => {
+    it('isStringArray', () => {
+      expect(isStringArray(['string', 'test'])).toEqual(true)
+      expect(isStringArray(['string', null])).toEqual(false)
+      expect(isStringArray(['string', undefined])).toEqual(false)
+      expect(isStringArray([{}, 'test'])).toEqual(false)
+      expect(isStringArray(['string', []])).toEqual(false)
+      expect(isStringArray([[]])).toEqual(false)
+      expect(isStringArray([])).toEqual(true)
+      expect(isStringArray(null)).toEqual(false)
+      expect(isStringArray(undefined)).toEqual(false)
+    })
+    it('isLocationArray', () => {
+      expect(isLocationArray()).toEqual(false)
+      expect(isLocationArray(null)).toEqual(false)
+      expect(isLocationArray(undefined)).toEqual(false)
+      expect(isLocationArray([])).toEqual(false)
+      expect(isLocationArray(['test'])).toEqual(false)
+      expect(isLocationArray([zipper(I.fromJS(data))])).toEqual(true)
+      expect(
+        isLocationArray([zipper(I.fromJS(data)), zipper(I.fromJS(data))])
+      ).toEqual(true)
+      expect(
+        isLocationArray([
+          zipper(I.fromJS(data)),
+          'Not a Zipper',
+          zipper(I.fromJS(data)),
+        ])
+      ).toEqual(false)
+    })
+  })
+
+  describe('predicates', () => {
```

Comment:
These aren't *predicates*. A **predicate** is a fn that takes a single argument and returns boolean (`predicate:: a -> Bool`)

Reply:
Thanks for the explanation of what predicates are. Mind-blowing really. 😜

The notion of predicates was overridden in my mind due to the description from https://github.com/netceteragroup/skele/issues/53:
> Each predicate can be:
> * a 'string' which is interpreted as descending into a property. The property should have an element (or child elements)
> * an array of strings which is interpreted as the predicate ofKind
> * a function returning a list of locations
> * a predicate function used to filter out nodes

... and the referenced documentation from [clojure.xml.zip](https://crossclj.info/ns/org.clojure/data.zip/0.1.2/clojure.data.zip.xml.html#_xml-%3E) where we have:
> The loc is passed to the first predicate.  If the predicate returns
> a collection, each value of the collection is passed to the next
> predicate.  If it returns a location, the location is passed to the
> next predicate.  If it returns true, the input location is passed to
> the next predicate.  If it returns false or nil, the next predicate
> is not called.

Shell I put `ofKind` as part of **predicates**, `child` as part of **motions** and the rest (`children`, `childrenFor`, `ancestors`, `descendants`) as **???** ?

Reply:
I've split the functions that can be used in `select` into:
* predicates
* motions
* selectors

Hope that this makes more sense and is clear